File: demo/app/android/app/src/main/kotlin/walletsdk/WalletSDK.kt
/*
Copyright Gen Digital Inc. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package walletsdk

import dev.trustbloc.wallet.sdk.api.Crypto
import dev.trustbloc.wallet.sdk.api.DIDDocResolution
import dev.trustbloc.wallet.sdk.api.DIDResolver
import dev.trustbloc.wallet.sdk.api.Header
import dev.trustbloc.wallet.sdk.did.Resolver
import dev.trustbloc.wallet.sdk.did.ResolverOpts
import dev.trustbloc.wallet.sdk.didion.Didion
import dev.trustbloc.wallet.sdk.didjwk.Didjwk
import dev.trustbloc.wallet.sdk.didkey.Didkey
import dev.trustbloc.wallet.sdk.display.Display
import dev.trustbloc.wallet.sdk.localkms.KMS
import dev.trustbloc.wallet.sdk.localkms.Localkms
import dev.trustbloc.wallet.sdk.localkms.Store
import dev.trustbloc.wallet.sdk.mem.ActivityLogger
import dev.trustbloc.wallet.sdk.verifiable.CredentialsArray
import walletsdk.openid4ci.OpenID4CI
import walletsdk.openid4ci.WalletInitiatedOpenID4CI
import walletsdk.openid4vp.OpenID4VP
import dev.trustbloc.wallet.sdk.attestation.Attestation
import dev.trustbloc.wallet.sdk.api.VerificationMethod

class WalletSDK {
    private var kms: KMS? = null
    var didResolver: DIDResolver? = null
    private var crypto: Crypto? = null
    var activityLogger: ActivityLogger? = null


    fun initSDK(kmsStore: Store, didResolverURI: String) {
        val kms = Localkms.newKMS(kmsStore)

        val opts = ResolverOpts()
        if (didResolverURI != "") {
            opts.setResolverServerURI(didResolverURI)
        }
        didResolver = Resolver(opts)

        crypto = kms.crypto
        activityLogger = ActivityLogger()
        this.kms = kms
    }

    fun createDID(didMethodType: String, didKeyType: String): DIDDocResolution {
        val kms = this.kms ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        val jwk = kms.create(didKeyType)

        println("Created a new key. The key ID is ${jwk.id()}")

        val doc: DIDDocResolution

        if (didMethodType == "key") {
            doc = Didkey.create(jwk)
        } else if (didMethodType == "jwk") {
            doc = Didjwk.create(jwk)
        } else if (didMethodType == "ion") {
            doc = Didion.createLongForm(jwk)
        } else {
            throw java.lang.Exception("DID method type $didMethodType not supported")
        }

        println("Successfully created a new did:${didMethodType} DID. The DID is ${doc.id()}")

        return doc

    }

    fun createOpenID4CIInteraction(requestURI: String): OpenID4CI {
        val didResolver = this.didResolver
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        val crypto = this.crypto
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        val activityLogger = this.activityLogger
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        val kms =
                this.kms
                        ?: throw java.lang.Exception("Local kms is not initialized, call initSDK()")
        return OpenID4CI(
                requestURI,
                crypto,
                didResolver,
                activityLogger,
                kms
        )
    }

    fun createOpenID4CIWalletInitiatedInteraction(issuerURI: String): WalletInitiatedOpenID4CI {
        val didResolver = this.didResolver
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        val crypto = this.crypto
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        return WalletInitiatedOpenID4CI(
                issuerURI,
                crypto,
                didResolver,
        )
    }


    fun createOpenID4VPInteraction(): OpenID4VP {
        val crypto = this.crypto
                ?: throw java.lang.Exception("crypto is not initialized, call initSDK()")
        val didResolver = this.didResolver
                ?: throw java.lang.Exception("did resolver is not initialized, call initSDK()")

        val activityLogger = this.activityLogger
                ?: throw java.lang.Exception("activity logger is not initialized, call initSDK()")

        return OpenID4VP(crypto, didResolver, activityLogger)
    }

    fun resolveCredentialsDisplayData(creds: CredentialsArray, issuerURI: String)
            : dev.trustbloc.wallet.sdk.display.Data {
        val didResolver = this.didResolver
                ?: throw java.lang.Exception("SDK is not initialized, call initSDK()")

        return Display.resolve(creds, issuerURI,
                dev.trustbloc.wallet.sdk.display.Opts().setDIDResolver(didResolver))
    }

    fun getAttestationVC(
            didVerificationMethod: VerificationMethod,
            attestationURL: String,
            disableTLSVerify: Boolean,
            attestationPayload: String,
            attestationToken: String?,
    ): String {
        val opts = Attestation.newCreateClientArgs(
                attestationURL,
                crypto,
        );
        if (disableTLSVerify) {
            opts.disableHTTPClientTLSVerify()
        }

        if (attestationToken != null) {
            opts.addHeader(Header("Authorization", "Bearer $attestationToken"))
        }

        val attestClient = Attestation.newClient(opts)

        val attestationVC = attestClient.getAttestationVC(
                didVerificationMethod,
                attestationPayload)

        return attestationVC.serialize()
    }
}

File: demo/app/ios/Runner/WalletSDK.swift
/*
 Copyright Gen Digital Inc. All Rights Reserved.

 SPDX-License-Identifier: Apache-2.0
 */

import Foundation
import Walletsdk

enum WalletSDKError: Error {
    case runtimeError(String)
}

class WalletSDK {
    private var kms:LocalkmsKMS?
    public var didResolver: ApiDIDResolverProtocol?
    private var crypto: ApiCryptoProtocol?
    var activityLogger: MemActivityLogger?



    func initSDK(kmsStore: LocalkmsStoreProtocol, didResolverURI: String) {
        kms = LocalkmsNewKMS(kmsStore, nil)

        let opts = DidNewResolverOpts()
        opts!.setResolverServerURI(didResolverURI)

        didResolver = DidNewResolver(opts, nil)
        crypto = kms!.getCrypto()
        activityLogger = MemActivityLogger()
    }

    func createDID(didMethodType: String, didKeyType: String) throws -> ApiDIDDocResolution {
        let jwk = try self.kms!.create(didKeyType)

        print("Created a new key. The key ID is \(jwk.id_())")

        var didCreateError: NSError?
        var doc: ApiDIDDocResolution?

        if didMethodType == "key" {
            doc = DidkeyCreate(jwk, &didCreateError)
        } else if didMethodType == "jwk" {
            doc = DidjwkCreate(jwk, &didCreateError)
        } else if didMethodType == "ion" {
            doc = DidionCreateLongForm(jwk, &didCreateError)
        } else {
            throw WalletSDKError.runtimeError("DID method type \(didMethodType) not supported")
        }

        if let didCreateError = didCreateError {
            throw didCreateError
        }

        var docIDError: NSError?

        let docID = doc!.id_(&docIDError)

        if let docIDError = docIDError {
            throw docIDError
        }

        print("Successfully created a new did:\(didMethodType) DID. The DID is \(docID)")

        return doc!
    }

    func createOpenID4CIInteraction(requestURI: String, correlationID: String) throws -> OpenID4CI {
        guard let crypto = self.crypto else {
            throw WalletSDKError.runtimeError("SDK is not initialized, call initSDK()")
        }
        guard let didResolver = self.didResolver else {
            throw WalletSDKError.runtimeError("SDK is not initialized, call initSDK()")
        }

        guard let localKMS = self.kms else {
            throw WalletSDKError.runtimeError("Local kms is not initialized, call initSDK()")
        }

        activityLogger = MemNewActivityLogger()


        return try OpenID4CI(requestURI: requestURI, didResolver: didResolver, crypto: crypto, activityLogger: activityLogger!, kms: localKMS, correlationID: correlationID)
    }

    func createOpenID4CIWalletInitiatedInteraction(issuerURI: String) throws -> WalletInitiatedOpenID4CI {
        guard let crypto = self.crypto else {
            throw WalletSDKError.runtimeError("SDK is not initialized, call initSDK()")
        }
        guard let didResolver = self.didResolver else {
            throw WalletSDKError.runtimeError("SDK is not initialized, call initSDK()")
        }

        return WalletInitiatedOpenID4CI(issuerURI: issuerURI, didResolver: didResolver, crypto: crypto)
    }

    func createOpenID4VPInteraction(correlationID: String) throws -> OpenID4VP {
        guard let crypto = self.crypto else {
            throw WalletSDKError.runtimeError("crypto is not initialized, call initSDK()")
        }
        guard let didResolver = self.didResolver else {
            throw WalletSDKError.runtimeError("did resolver is not initialized, call initSDK()")
        }
        guard let activityLogger = self.activityLogger else {
            throw WalletSDKError.runtimeError("activity logger is not initialized, call initSDK()")
        }
        guard let kms = self.kms else {
            throw WalletSDKError.runtimeError("kms is not initialized, call initSDK()")
        }
        return OpenID4VP(didResolver: didResolver, crypto: crypto, activityLogger: activityLogger, kms: kms, correlationID: correlationID)
    }

    func getAttestationVC(
        didVerificationMethod: ApiVerificationMethod,
        attestationURL: String,
        disableTLSVerify: Bool,
        attestationPayload: String,
        attestationToken: String?
   ) throws -> String {
        let opts = AttestationNewCreateClientArgs(attestationURL, crypto)!
        if (disableTLSVerify) {
            opts.disableHTTPClientTLSVerify()
        }
        if (attestationToken != nil) {
            opts.add(ApiHeader("Authorization", value: "Bearer "+attestationToken!))
        }

        // opts.add(ApiHeader("X-Correlation-Id", value: "ramdomstr"))

        let attestClient = AttestationNewClient(opts, nil)!


        let attestationVC = try attestClient.getAttestationVC(didVerificationMethod, attestationPayloadJSON: attestationPayload)

        return try attestationVC.serialize(nil)
    }
}

